utilities/include/vec4.hpp

```diff
@@ -0,0 +1,151 @@
+#ifndef VEC4_HPP
+#define VEC4_HPP
+
+#include <cstddef>
+#include <ostream>
+#include <cmath>
+#include <xsimd/xsimd.hpp>
+
+#include "config.hpp"
+
+class alignas(ALIGN_WIDTH) Vec4{
+    
+    public:
+        constexpr Vec4() : m_v{0,0,0,0} {}
+        constexpr explicit Vec4(Utype x) : m_v{x,x,x,x} {}
+        constexpr Vec4(Utype x,Utype y,Utype z,Utype w) : m_v{x,y,z,w} {}
+        constexpr explicit Vec4(xsimd::batch<Utype,UArch> x): m_v{0,0,0,0} {x.store_aligned(m_v);}
+        
+        //getters
+        constexpr Utype x() const{
+            return m_v[0];
+        }
+
+        constexpr Utype y() const{
+            return m_v[1];
+        }
+        
+        constexpr Utype z() const{
+            return m_v[2];
+        }
+
+        constexpr Utype w() const{
+            return m_v[3];
+        }
+        
+        //indexing operations
+        // constexpr Utype& operator[](std::size_t i) const{
```

Comment:
My bad, I had the wrong signatures in my before. The two function signatures you want are `constexpr UType& operator[](const std::size_t i)`, and `constexpr const UType& operator[](const std::size_t i) const`.

When you are making a reference to a type `constexpr`, you also need the `const` because that tells the compiler that the _reference_ is also `const`, which allows r-value to const l-value conversions. (You need an l-value to take a reference.)